File: app/src/UI/Features/Records/Activity/forms/common/validators/noRepeatKey.ts
/**
 * @desc Iterate an Array of Objects searching if any key matches another. e.g.: Two entries have the same invasive plant
 * @param arr Array of entries
 * @param key Key to compare e.g. 'invasive_plant'
 * @param keyLabel Plain language representation of key to appear in error message e.g. 'Invasive Plant'
 */
const noRepeatKey = <T extends Record<PropertyKey, unknown>>(
  arr: T[],
  key: keyof T,
  keyLabel?: string
): boolean | string => {
  const seen = new Set();

  for (const entry of arr) {
    const value = entry?.[key];
    // Skip empty fields
    if (value === undefined || value === null || value === '') continue;
    if (seen.has(value)) {
      return `The same ${keyLabel ?? String(key)} cannot appear in multiple entries.`;
    }
    seen.add(value);
  }
  return true;
};

export default noRepeatKey;

File: api-rework/invasives/api/protocol/activity/validators/no_repeat_key.py
from typing import Iterable, Optional, TypeVar

T = TypeVar("T")


def no_repeat_key(
    entries: Iterable[T], key: str, key_label: Optional[str] = None
) -> Iterable[T]:
    """
    Generic logic to ensure no duplicate values exist for a specific key
    within a collection of objects.
    """
    seen = set()
    label = key_label or key

    for entry in entries:
        value = getattr(entry, key, None)
        if value in (None, ""):
            continue
        if value in seen:
            raise ValueError(f"The same {label} cannot appear in multiple entries.")
        seen.add(value)
    return entries
